src/classes/RD2_DataMigration_BATCH.cls

```diff
@@ -114,40 +167,125 @@ public class RD2_DataMigration_BATCH implements Database.Batchable<SObject>, Sch
+    * @return void
     */
-    private void handleDmlAndLogging(Savepoint sp) {
-        // TODO Rewrite this to capture and log dml exceptions in a new logging service
-        //Database.update(convertedRDs, false);
-        update convertedRDs;
+    private void handleSaveAndLogging(
+        List<npe03__Recurring_Donation__c> convertedRDs, 
+        Map<Id, List<Opportunity>> voidOppsByRDId,
+        Logger logger
+    ) {
+        Savepoint sp = Database.setSavepoint();
+
+        Set<Id> savedRDIds = saveRecurringDonations(convertedRDs, logger);
+
+        voidOpenOpportunities(savedRDIds, voidOppsByRDId, logger);
 
         if (this.isDryRunMode) {
             Database.rollback(sp);
         }
 
-        // TODO Commit any logging records after the rollback
+        logger.processErrors();
+    }
+
+    /**
+    * @description Save Recurring Donations
+    * @param convertedRDs Converted Recurring Donations
+    * @param logger Exception/error log container
+    * @return Set<Id> Successfully saved Recurring Donation Ids
+    */
+    private Set<Id> saveRecurringDonations(List<npe03__Recurring_Donation__c> convertedRDs,  Logger logger) {
+
+        List<Database.SaveResult> saveResults = convertedRDs.isEmpty()
+            ? new List<Database.SaveResult>()
+            : Database.update(convertedRDs, false);
+        
+        logger.addErrors(saveResults, convertedRDs);
+
+        return getSavedRecordIds(saveResults);
+    }
+
+    /**
+    * @description Returns Ids of records that were saved successfully
+    * @param saveResults Insert/update DML save results
+    * @return Set<Id>
+    */
+    private Set<Id> getSavedRecordIds(List<Database.SaveResult> saveResults) {
+        Set<Id> recordIds = new Set<Id>();
+
+        for (Database.SaveResult result : saveResults) {
+            if (result.isSuccess()) {
+                recordIds.add(result.getId());
+            }
+        }
+
+        return recordIds;
+    }
+
+    /**
+    * @description Handles Open Opportunities that should be voided (sets them to Closed Lost or deletes them)
+    * @param savedRDIds Ids of successfully saved Recurring Donations
+    * @param voidOppsByRDId To be voided Open Opportunities by Recurring Donation Id
+    * @param logger Exception/error log container
+    * @return void
+    */
+    private void voidOpenOpportunities(Set<Id> savedRDIds, Map<Id, List<Opportunity>> voidOppsByRDId, Logger logger) {
+        List<Opportunity> opps = new List<Opportunity>();
+        for (Id rdId : savedRDIds) {
+            if (voidOppsByRDId.containsKey(rdId)) {
+                opps.addAll(voidOppsByRDId.get(rdId));
+            }
+        }
+
+        if (opps.isEmpty()) {
+            return;
+        }
+
+        List<Object> dmlResults = new List<Object>();
+        if (rdSettings.npe03__Open_Opportunity_Behavior__c == RD2_Constants.CloseActions.Delete_Open_Opportunities.name()) {
+            dmlResults = Database.delete(opps, false);
```

Comment:
@force2b , if an Open Opp deletion or update to Closed Lost fails, Recurring Donation will still be migrated and an error log will be created for the failed Opp. Is this expected behaviour?
FYI @Julian88Tex

Reply:
@force2b , as discussed, I will make the change to rollback the whole batch iteration if Opp deletion/update fails.

Reply:
@force2b , batch iteration rollback change has been applied.

Reply:
Thanks @RadSFDeveloper. Ideally once we move to a _successful-dry-run-required_ enablement process this won't matter as much.